resolvers/itemResolver.js

```diff
@@ -0,0 +1,160 @@
+module.exports = {
+    Query: {
+        item(obj, args, context, info) {
+            if (args.id) return context.data.item.getItem(args.id);
+            if (args.normalizedName) return context.data.item.getItemByNormalizedName(args.normalizedName);
+            return Promise.reject(new Error('You must specify either the id or the normalizedName argument'));
+        },
+        async items(obj, args, context) {
+            let items = false;
+            let arguments = {
+                ids: {
+                    query: async ids => {
+                        return context.data.item.getItemsByIDs(ids);
+                    },
+                    filter: ids => {
+                        return items.filter(item => {
+                            return ids.includes[item.id];
+                        });
+                    }
+                },
+                name: {
+                    query: async name => {
+                        return context.data.item.getItemsByName(name);
+                    },
+                    filter: name => {
+                        const searchString = name.toLowerCase();
+                        return items.filter(item => {
+                            return item.name.toLowerCase().includes(searchString) || item.shortname.toLowerCase().includes(searchString);
+                        });
+                    }
+                },
+                type: {
+                    query: async type => {
+                        return context.data.item.getItemsByType(type);
+                    },
+                    filter: type => {
+                        return items.filter(item => {
+                            return item.types.includes(type);
+                        });
+                    }
+                },
+                bsgCategoryId: {
+                    query: async bsgcat => {
+                        return context.data.item.getItemsByBsgCategoryId(bsgcat);
+                    },
+                    filter: bsgcat => {
+                        return items.filter(item => {
+                            return item.bsgCategoryId === bsgcat;
+                        });
+                    }
+                },
+            }
+            if (Object.keys(args).length === 0) return context.data.item.getAllItems();
+            for (const argName in args) {
+                if (!arguments[argName]) return Promise.reject(new Error(`${argName} is not a recognized argument`));
+                if (!items) {
+                    items = await arguments[argName].query(args[argName]);
+                } else {
+                    items = arguments[argName].filter(args[argName]);
+                }
+            }
+            return items;
+        },
+        itemsByIDs(obj, args, context, info) {
```

Comment:
I see in many of these functions and the ones below it, `obj` is a param but it is not used in the function. Is this intended?

Reply:
Yes, it's necessary to specify the obj param to get to the args and context params. I also specify the unused "info" param in lots of places just to remind myself that it's there. We could remove that one, but it's not hurting anything.